### PR TITLE
#3883 Cannot import products 

### PR DESCRIPTION
**Pull request by cristinamghita**
#3883 Cannot import products

### CODE REVIEW

File: de.metas.business/src/main/java/org/adempiere/impexp/product/MProductImportTableSqlUpdater.java [hunk 1]
```diff
@@ -414,7 +414,7 @@ private void dbUpdateTaxCategories(@NonNull final String whereClause, @NonNull f
 		{
 			sql = new StringBuilder("UPDATE ")
 					.append(targetTableName + " i ")
-					.append(" set C_TaxCategory_ID=(select tc.C_TaxCategory_ID from C_TaxCategory tc where tc.IsDefault='Y' and tc.AD_Client_ID=")
+					.append(" set C_TaxCategory_ID=(select tc.C_TaxCategory_ID from C_TaxCategory tc where tc.VATType='N' and tc.AD_Client_ID=")
```
Comment (teosarca) [inline de.metas.business/src/main/java/org/adempiere/impexp/product/MProductImportTableSqlUpdater.java]: why not trying to fetch the C_TaxCategory_ID by using some API, and then here, just simple set it.
In that way we would prevent errors like the one u've fixed here.

File: de.metas.business/src/main/java/org/adempiere/impexp/product/MProductImportTableSqlUpdater.java [hunk 2]
```diff
@@ -562,5 +562,17 @@ private void dbUpdateErrorMessages(@NonNull final String whereClause)
...
+
+
+		sql = new StringBuilder("UPDATE ")
+				.append(targetTableName + " i ")
+				.append(" SET " + COLUMNNAME_I_IsImported + "='E', " + COLUMNNAME_I_ErrorMsg + "=" + COLUMNNAME_I_ErrorMsg + "||'ERR=No mandatory Pharma product category Name,' ")
+				.append("WHERE PharmaProductCategory_Name IS NOT NULL ")
+				.append("AND M_PharmaProductCategory_ID IS NULL")
+				.append(" AND " + COLUMNNAME_I_IsImported + "<>'Y'").append(whereClause);
+		no = DB.executeUpdateEx(sql.toString(), ITrx.TRXNAME_ThreadInherited);
+		logger.warn("No Mandatory Pharma product category Name =" + no);
```
Comment (teosarca) [inline de.metas.business/src/main/java/org/adempiere/impexp/product/MProductImportTableSqlUpdater.java]: no string concatenation pls

File: de.metas.vertical.pharma/src/main/java/org/adempiere/impexp/impl/PharmaImportProductInterceptor.java [hunk 3]
```diff
@@ -101,6 +108,11 @@ public void onImport(IImportProcess<?> process, Object importModel, Object targe
 		importPrices(iproduct);
 	}
 
+	private boolean isPharmaProductImport(final I_I_Product iproduct)
+	{
+		return !Check.isEmpty(iproduct.getPharmaProductCategory_Name(), true);
+	}
```
Comment (teosarca) [inline de.metas.vertical.pharma/src/main/java/org/adempiere/impexp/impl/PharmaImportProductInterceptor.java]: why `pharma product import` is considered true when `PharmaProductCategory_Name` is set?

Comment (teosarca) [reply de.metas.vertical.pharma/src/main/java/org/adempiere/impexp/impl/PharmaImportProductInterceptor.java]: and more, what means `isPharmaProductImport`?

**Comment by teosarca** [comment]
pls check my comments

File: de.metas.vertical.pharma/src/main/java/org/adempiere/impexp/impl/PharmaImportProductInterceptor.java (hunk 3 above)
Comment (cristinamghita) [reply de.metas.vertical.pharma/src/main/java/org/adempiere/impexp/impl/PharmaImportProductInterceptor.java]: We have some special pharma fields, that in the regular product imports don't exist, but could crash the standard product import.
Basically, I don't want that interceptor' code,PharmaImportProductInterceptor, to run, if is a standard import product, with no pharma fields.

Comment (teosarca) [reply de.metas.vertical.pharma/src/main/java/org/adempiere/impexp/impl/PharmaImportProductInterceptor.java]: got it.
But then, for clarity, i would suggest to move that code from line ~80 to 108 to a method called `onPharmaProductImported`. Call that method if `isPharmaProductImport(iproduct)`.

File: de.metas.swat/de.metas.swat.base/src/main/java/de/metas/tax/api/impl/TaxBL.java [hunk 4]
```diff
@@ -561,4 +563,17 @@ public void setupIfIsWholeTax(final I_C_Tax tax)
 		tax.setIsDocumentLevel(true);
 		// tax.setIsSalesTax(false); // does not matter
 	}
+
+	@Override
+	public int retrieveRegularTaxCategoryId()
```
Comment (teosarca) [inline de.metas.swat/de.metas.swat.base/src/main/java/de/metas/tax/api/impl/TaxBL.java]: what shall happen when there is no tax category found?

Comment (teosarca) [reply de.metas.swat/de.metas.swat.base/src/main/java/de/metas/tax/api/impl/TaxBL.java]: and... why not caching it?

**Pull request merged**